### ISSUE TITLE
Implement Basic Linting

### ISSUE THREAD

**Issue by kyorav** [np-guard/vpc-network-config-analyzer#116]
This issue is for brainstorming ideas for checks we may include as part of a linting functionality. At some point we will review the list with a stakeholder and decide which ones make sense.

1. partial connections. When a request can be sent to destination but the response will be blocked. #706 solves
1. nACL splits a subnet. When a cidr that does not cover the entire subnet is used in the source of an outbound rule or the destination of an inbound rule. #698 solves
2. sg split subnet as above. #698 solves
1. LB analysis: warn about SG that split subnet of a configured LB (without running full connectivity analysis)
4. As the above two items, but when there is an actual connectivity gap between two endpoints in the same subnet [define: what is considered an endpoint]
1. overlapping subnet ranges (relevant mostly for the multiple VPCs use case) #700 solves
1. source-port limitation may block response. This will help identify misuse of source-port. Note that this is relevant when the connection is complete. (@kyorav please elaborate) 
1. Non-connected tables: #722 solves
    1. SG not connected to an endpoint
    2. NACL not connected to a subnet
1. Non-connected Routers
    1. tgw not connected
    2. fip not connnected
    3. pgw not connected
    4. loadbalancer without pool members
1. Redundant Subnets (no connected endpoints)
2. Redundant VPC....
1. Redundant rules:
    1. SG/NACL rule that is useless w.r.t. the relevant address prefixes #115 solved by #731
    1. SG rule that is implied by other SG rules in the same table
    1. NACL rule that is implied by other, higher priories, NACL rules in the same table
    1. SG/NACL rules that do not actually effect the connectivity    
1. Routing Tables analysis: warn about non deterministic routing 
1. NACL rules with values in src/dst min/max port but no value in protocol - in which case the ports are ignored (discuss)

**Comment by adisos** [comment]
Related: issue#115

**Comment by ShiriMoran** [comment]
> This issue is for brainstorming ideas for checks we may include as part of a linting functionality. At some point we will review the list with a stakeholder and decide which ones make sense.
> 
>     * partial connections. When a request can be sent to destination but the response will be blocked.
> 
>     * nACL splits a subnet. When a cidr that does not cover the entire subnet is used in the source of an outbound rule or the destination of an inbound rule.
> 
>     * overlapping subnet ranges (relevant mostly for the multiple VPCs use case)
> 
>     * source-port limitation may block response. This will help identify misuse of source-port. Note that this is relevant when the connection is complete.
> 
>     * Related: issue [Add a warning for useless sg rules based on the VPC APs #115](https://github.com/np-guard/vpc-network-config-analyzer/issues/115)
> 
>     * LB analysis: warn about SG that split subnet of a configured LB (without running full connectivity analysis)
Only SG are relevant in this regard? what about s.g. prefix filters? Perhaps rephrase:  
** warn when different (random) selections of private IPs may yield different connectivity results for LB

> 
>     * Routing Tables analysis: warn about non deterministic routing

**Comment by kyorav** [comment]
> warn when different (random) selections of private IPs may yield different connectivity results for LB
This is not linting, you need to apply the complete analysis to figure this out. And yes, our analysis will highlight this.

**Comment by zivnevo** [comment]
> > warn when different (random) selections of private IPs may yield different connectivity results for LB
> 
> This is not linting, you need to apply the complete analysis to figure this out. And yes, our analysis will highlight this.

I'm not sure we want to limit linting to just shallow analysis. Wikipedia says:
```
Lint is the computer science term for a static code analysis tool used to flag programming errors, bugs, stylistic errors and suspicious constructs.
```
so I think I would like to include everything suspicious we can find.

**Comment by kyorav** [comment]
>so I think I would like to include everything suspicious we can find

What I mean is that our analysis functionality checks this (by inserting "fake" LB endpoints in all subnets and notifying the user when they don't all have the same connectivity) so I don't think we need to add it to the linting functionality. Or, at least, we don't need to implement it as part of this issue because it is already implemented. Unless you somehow want to implement it differently for the linting functionality?

**Comment by zivnevo** [comment]
It makes sense for both the analysis (`vpcanalyzer report`) and the linter (`vpcanalyzer lint`) to notify the user about such issues, as these subcommands are used for different use-cases.
We should make sure the code for this check is not duplicated.

**Comment by ShiriMoran** [comment]
>     * nACL splits a subnet. When a cidr that does not cover the entire subnet is used in the source of an outbound rule or the destination of an inbound rule.

What about (local/remote cidr blocks splitting subnets in rules of) SG? @kyorav

**Comment by kyorav** [comment]
>What about (local/remote cidr blocks splitting subnets in rules of) SG

This is a weaker indication of error, but can still fit as a linting rule. A common correct pattern is to have a security group rule that allows access to the address of a load balancer, DNS, proxy, etc. So if the cidr size is a single address ("/32") we probably don't want to flag it, so as not to create too much noise.

**Comment by ShiriMoran** [comment]
First demo: 
* 2,3,6 (merged or awaits PR)
* todo: 1, 8, 14, 12 i-iii

**Comment by ShiriMoran** [comment]
1 solved by PR #706

**Comment by ShiriMoran** [comment]
followup #800
